controllers/dns_workflow.go

```diff
@@ -1,7 +1,141 @@
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnspolicies/finalizers,verbs=update
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnsrecords,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnsrecords/status,verbs=get
+
+func NewDNSWorkflow(client *dynamic.DynamicClient, scheme *runtime.Scheme) *controller.Workflow {
+	return &controller.Workflow{
+		Precondition: NewDNSPoliciesValidator().Subscription().Reconcile,
+		Tasks: []controller.ReconcileFunc{
+			NewEffectiveDNSPoliciesReconciler(client, scheme).Subscription().Reconcile,
+		},
+		Postcondition: NewDNSPolicyStatusUpdater(client).Subscription().Reconcile,
+	}
+}
+
+func LinkListenerToDNSRecord(objs controller.Store) machinery.LinkFunc {
+	gateways := lo.Map(objs.FilterByGroupKind(machinery.GatewayGroupKind), controller.ObjectAs[*gwapiv1.Gateway])
+	listeners := lo.FlatMap(lo.Map(gateways, func(g *gwapiv1.Gateway, _ int) *machinery.Gateway {
+		return &machinery.Gateway{Gateway: g}
+	}), machinery.ListenersFromGatewayFunc)
+
+	return machinery.LinkFunc{
+		From: machinery.ListenerGroupKind,
+		To:   DNSRecordGroupKind,
+		Func: func(child machinery.Object) []machinery.Object {
+			return lo.FilterMap(listeners, func(l *machinery.Listener, _ int) (machinery.Object, bool) {
+				if dnsRecord, ok := child.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord); ok {
+					return l, l.GetNamespace() == dnsRecord.GetNamespace() &&
+						dnsRecord.GetName() == dnsRecordName(l.Gateway.Name, string(l.Name))
+				}
+				return nil, false
+			})
+		},
+	}
+}
+
+func LinkDNSPolicyToDNSRecord(objs controller.Store) machinery.LinkFunc {
+	policies := lo.Map(objs.FilterByGroupKind(v1alpha1.DNSPolicyGroupKind), controller.ObjectAs[*v1alpha1.DNSPolicy])
+
+	return machinery.LinkFunc{
+		From: v1alpha1.DNSPolicyGroupKind,
+		To:   DNSRecordGroupKind,
+		Func: func(child machinery.Object) []machinery.Object {
+			if dnsRecord, ok := child.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord); ok {
+				return lo.FilterMap(policies, func(dnsPolicy *v1alpha1.DNSPolicy, _ int) (machinery.Object, bool) {
+					return dnsPolicy, utils.IsOwnedBy(dnsRecord, dnsPolicy)
+				})
+			}
+			return nil
+		},
+	}
+}
+
+func dnsPolicyAcceptedStatusFunc(state *sync.Map) func(policy machinery.Policy) (bool, error) {
+	validatedPolicies, validated := state.Load(StateDNSPolicyAcceptedKey)
+	if !validated {
+		return dnsPolicyAcceptedStatus
+	}
+	validatedPoliciesMap := validatedPolicies.(map[string]error)
+	return func(policy machinery.Policy) (bool, error) {
+		err, pValidated := validatedPoliciesMap[policy.GetLocator()]
+		if pValidated {
+			return err == nil, err
+		}
+		return dnsPolicyAcceptedStatus(policy)
+	}
+}
+
+func dnsPolicyAcceptedStatus(policy machinery.Policy) (accepted bool, err error) {
+	p, ok := policy.(*v1alpha1.DNSPolicy)
+	if !ok {
+		return
```

Comment:
is it expected we will get non DNSPolicies? or should there be an error returned?

Reply:
If you passed on in you would, but where its being used it would never not be. That said, we don't want to return an error in this case as the errors returned from this function are supposed to be related to the dnspolicy status.

Same code as RLP for consistency https://github.com/Kuadrant/kuadrant-operator/blob/main/controllers/ratelimit_workflow.go#L267-L280